### PR TITLE
Converts worker watcher to use django models and removes ScheduledCalls

### PR DESCRIPTION
**Pull request by dkliban**
This also renames the scheduler to pulp_celerybeat.

### CODE REVIEW

**Comment by BrnoPCmaniak** [comment]
Just an suggestion on scheduler https://github.com/BrnoPCmaniak/pulp/commit/f8bd4cb12020c8adf0f616f77dcdc756c097a7b7?diff=split

**Comment by dkliban** [comment]
@BrnoPCmaniak Thanks for the suggestion. I'll update my PR with your changes. 

I am also going to remove the ScheduledCall references because we will no longer have user defined scheduled tasks stored in Pulp.

File: tasking/pulp/tasking/worker_watcher.py [hunk 1]
```diff
@@ -72,14 +72,16 @@ def handle_worker_heartbeat(event):
     :type event: dict
     """
     event_info = _parse_and_log_event(event)
-    worker = Worker.objects(name=event_info['worker_name']).first()
 
-    if not worker:
+    try:
+        worker = Worker.objects.get(name=event_info['worker_name'])
+    except worker.DoesNotExist:
+        worker = Worker(name=event_info['worker_name'])
```
Comment (BrnoPCmaniak) [inline tasking/pulp/tasking/worker_watcher.py]: I think this would be better using [`worker, created = Worker.objects.get_or_create(name=event_info['worker_name'])`](https://docs.djangoproject.com/en/1.10/ref/models/querysets/#get-or-create)

File: tasking/pulp/tasking/scheduler.py [hunk 2]
```diff
@@ -117,38 +114,39 @@ def check_celery_processes(self):
...
+            _logger.error(msg)
+
+            if worker.name.startswith(constants.SCHEDULER_WORKER_NAME):
+                worker.delete()
             else:
-                worker_count = worker_count + 1
+                _delete_worker(worker.name)
 
-        if resource_manager_count == 0:
+        if not (Worker.objects.filter(name__startswith=constants.RESOURCE_MANAGER_WORKER_NAME)
```
Comment (dkliban) [inline tasking/pulp/tasking/scheduler.py]: I rearranged things and now just use the count variable that used to be declared just after this code.

File: tasking/pulp/tasking/scheduler.py [hunk 3]
```diff
@@ -117,38 +114,39 @@ def check_celery_processes(self):
...
+                      .exists()):
             msg = _("There are 0 pulp_resource_manager processes running. Pulp will not operate "
                     "correctly without at least one pulp_resource_mananger process running.")
             _logger.error(msg)
 
-        if scheduler_count == 0:
+        if not Worker.objects.filter(name__startswith=constants.SCHEDULER_WORKER_NAME).exists():
```
Comment (dkliban) [inline tasking/pulp/tasking/scheduler.py]: I did the same thing as above. Using the scheduler_count to do the check.

File: tasking/pulp/tasking/scheduler.py [hunk 4]
```diff
@@ -173,15 +171,12 @@ class Scheduler(beat.Scheduler):
     second, is a WorkerTimeoutMonitor thread that watches for cases where all workers disappear
     at once.
     """
-    Entry = ScheduleEntry
+    Entry = beat.ScheduleEntry
```
Comment (dkliban) [inline tasking/pulp/tasking/scheduler.py]: I had the same suspicion also. I am removing this line.

File: tasking/pulp/tasking/scheduler.py [hunk 5]
```diff
@@ -260,110 +255,34 @@ def tick(self):
         old_timestamp = datetime.utcnow() - timedelta(seconds=constants.CELERYBEAT_LOCK_MAX_AGE)
 
         # Updating the current lock if lock is on this instance of celerybeat
-        result = CeleryBeatLock.objects(celerybeat_name=self.celerybeat_name).\
-            update(set__timestamp=datetime.utcnow())
-
-        # If current instance has lock and updated lock_timestamp, call super
-        if result == 1:
+        try:
+            celerybeat_lock = TaskLock.objects.get(name=celerybeat_name)
```
Comment (dkliban) [inline tasking/pulp/tasking/scheduler.py]: this should be self.celerybeat_name which is defined in the __init__(). Changing it to that.

File: tasking/pulp/tasking/scheduler.py [hunk 6]
```diff
@@ -260,110 +255,34 @@ def tick(self):
...
+            celerybeat_lock.save()
+            # If current instance has lock and updated lock_timestamp, call super
             _logger.debug(_('Lock updated by %(celerybeat_name)s')
-                          % {'celerybeat_name': self.celerybeat_name})
-            ret = self.call_tick(self, self.celerybeat_name)
-        else:
-            # check for old enough time_stamp and remove if such lock is present
-            CeleryBeatLock.objects(timestamp__lte=old_timestamp).delete()
+                          % {'celerybeat_name': celerybeat_name})
+            ret = self.call_tick(self, celerybeat_name)
```
Comment (dkliban) [inline tasking/pulp/tasking/scheduler.py]: same

File: tasking/pulp/tasking/scheduler.py [hunk 7]
```diff
@@ -260,110 +255,34 @@ def tick(self):
...
+        except TaskLock.DoesNotExist:
+            TaskLock.objects(timestamp__lte=old_timestamp).delete()
```
Comment (dkliban) [inline tasking/pulp/tasking/scheduler.py]: good catch! adding the name filter here.

File: tasking/pulp/tasking/scheduler.py [hunk 8]
```diff
@@ -7,20 +8,15 @@
...
-from pulp.server.db.connection import UnsafeRetry
-from pulp.server.db.model import CeleryBeatLock, Worker
-from pulp.server.db.model.dispatch import ScheduledCall, ScheduleEntry
-from pulp.server.managers.schedule import utils
-from pulp.tasking import delete_worker
+from pulp.tasking import worker_watcher
 from pulp.tasking.celery_instance import celery as app
-from pulp.tasking.constants import TASKING_CONSTANTS
+from pulp.tasking import delete_worker
+from pulp.app.models.task import TaskLock, Worker
```
Comment (bmbouter) [inline tasking/pulp/tasking/scheduler.py]: This import is out of order

File: tasking/pulp/tasking/scheduler.py [hunk 9]
```diff
@@ -260,110 +251,34 @@ def tick(self):
...
+            celerybeat_lock.save()
+            # If current instance has lock and updated lock_timestamp, call super
             _logger.debug(_('Lock updated by %(celerybeat_name)s')
                           % {'celerybeat_name': self.celerybeat_name})
             ret = self.call_tick(self, self.celerybeat_name)
-        else:
-            # check for old enough time_stamp and remove if such lock is present
-            CeleryBeatLock.objects(timestamp__lte=old_timestamp).delete()
+        except TaskLock.DoesNotExist:
+            TaskLock.objects(name=self.celerybeat_name, timestamp__lte=old_timestamp).delete()
```
Comment (bmbouter) [inline tasking/pulp/tasking/scheduler.py]: This needs to be `TaskLock.objects.get` I think

File: tasking/pulp/tasking/scheduler.py [hunk 10]
```diff
@@ -260,110 +251,34 @@ def tick(self):
...
             try:
-                lock_timestamp = datetime.utcnow()
-
+                lock_timestamp = timezone.now()
                 # Insert new lock entry
-                new_lock = CeleryBeatLock(celerybeat_name=self.celerybeat_name,
-                                          timestamp=lock_timestamp)
-                new_lock.save()
+                TaskLock.objects.create(name=self.celerybeat_name, timestamp=lock_timestamp,
```
Comment (bmbouter) [inline tasking/pulp/tasking/scheduler.py]: The `timestamp` will be set to now() automatically so it's not needed and it can be removed from L#265 too.

File: tasking/pulp/tasking/scheduler.py [hunk 11]
```diff
@@ -117,38 +113,38 @@ def check_celery_processes(self):
...
+        resource_manager_count = Worker.objects.filter(
+            name__startswith=constants.RESOURCE_MANAGER_WORKER_NAME).count()
+
+        if not resource_manager_count:
             msg = _("There are 0 pulp_resource_manager processes running. Pulp will not operate "
                     "correctly without at least one pulp_resource_mananger process running.")
             _logger.error(msg)
 
-        if scheduler_count == 0:
+        if not scheduler_count:
```
Comment (bmbouter) [inline tasking/pulp/tasking/scheduler.py]: For readability I think this was clearer with the ` == 0`. It would be good to make them all consistent.

File: tasking/pulp/tasking/constants.py [hunk 12]
```diff
@@ -4,6 +4,20 @@
     # The interval in seconds for which the Celery Process monitor thread sleeps between
     # checking for missing Celery processes.
     CELERY_CHECK_INTERVAL=60,
+    # The interval in seconds during which the Celery Processes being monitored need to report a
+    # hearbeat to be considered active.
+    CELERY_TIMEOUT_SECONDS=200,
```
Comment (bmbouter) [inline tasking/pulp/tasking/constants.py]: Was this set to 200 before? I remember the age-out time being 300 seconds. Generally I want to lower these timing for Pulp3 anyway. Should we do that now or in another PR? I'm OK either way.

File: tasking/pulp/tasking/constants.py [hunk 13]
```diff
@@ -4,6 +4,20 @@
...
+    # The interval in seconds with which a secondary resource manager tries to acquire the lock
+    # that is needed to be the primary resource manager.
+    CELERY_TICK_DEFAULT_WAIT_TIME=90,
```
Comment (bmbouter) [inline tasking/pulp/tasking/constants.py]: Consider renaming this to be `RESOURCE_MANAGER_LOCK_RETRY_ACQUIRE_TIME` or `RESOURCE_MANAGER_LOCK_RETRY_TIME`

File: tasking/pulp/tasking/constants.py [hunk 14]
```diff
@@ -4,6 +4,20 @@
...
     # Resource manager worker name
-    RESOURCE_MANAGER_WORKER_NAME='resource_manager'
+    RESOURCE_MANAGER_WORKER_NAME='resource_manager',
+    # celerybeat constants
+    SCHEDULER_WORKER_NAME = 'pulp_celerybeat',
+    # The maximum amount of time in seconds celerybeat sleeps before checking for more scheduled
+    # work to dispatch.
+    CELERY_MAX_INTERVAL=90,
```
Comment (bmbouter) [inline tasking/pulp/tasking/constants.py]: s/CELERY_MAX_INTERVAL/CELERYBEAT_MAX_SLEEP_INTERVAL/
OR
s/CELERY_MAX_INTERVAL/CELERYBEAT_MAX_SLEEP/

File: tasking/pulp/tasking/constants.py [hunk 15]
```diff
@@ -4,6 +4,20 @@
...
+    # The amount of time in seconds that a celerybeat lock is valid for. If the lock is older than
+    # the value specified here, another instance of pulp_celerybeat is allowed to acquire the lock.
```
Comment (bmbouter) [inline tasking/pulp/tasking/constants.py]: s/another instance of pulp_celerybeat is allowed to acquire the lock/the process holding the lock is considered to be dead/

File: tasking/pulp/tasking/constants.py [hunk 16]
```diff
@@ -4,6 +4,20 @@
...
+    # hearbeat to be considered active.
+    CELERY_TIMEOUT_SECONDS=200,
+    # The interval in seconds with which a secondary resource manager tries to acquire the lock
+    # that is needed to be the primary resource manager.
+    CELERY_TICK_DEFAULT_WAIT_TIME=90,
     # Resource manager worker name
-    RESOURCE_MANAGER_WORKER_NAME='resource_manager'
+    RESOURCE_MANAGER_WORKER_NAME='resource_manager',
+    # celerybeat constants
+    SCHEDULER_WORKER_NAME = 'pulp_celerybeat',
```
Comment (dkliban) [inline tasking/pulp/tasking/constants.py]: sounds good to me. I am updating my PR wit this change.

File: tasking/pulp/tasking/constants.py (hunk 14 above)
Comment (dkliban) [inline tasking/pulp/tasking/constants.py]: i am going with `CELERYBEAT_MAX_SLEEP_INTERVAL`

File: tasking/pulp/tasking/constants.py (hunk 12 above)
Comment (dkliban) [inline tasking/pulp/tasking/constants.py]: i like `HEARTBEAT_MAX_AGE`.

File: tasking/pulp/tasking/constants.py (hunk 15 above)
Comment (dkliban) [inline tasking/pulp/tasking/constants.py]: thanks! i am updating the PR with this change.

File: tasking/pulp/tasking/constants.py (hunk 13 above)
Comment (dkliban) [inline tasking/pulp/tasking/constants.py]: So I looked at the code, and the comment was wrong. This value is used by the celerybeat only. It's how long celerybeat sleeps between trying to acquire the lock. I am changing this to `CELERYBEAT_LOCK_RETRY_TIME` and updating the comment.

File: tasking/pulp/tasking/constants.py [hunk 17]
```diff
@@ -4,6 +4,20 @@
...
+    # The interval in seconds during which the Celery Processes being monitored need to report a
+    # hearbeat to be considered active.
+    HEARTBEAT_MAX_AGE=200,
+    # The interval in seconds with which a secondary celerybeat tries to acquire the lock
+    # that is needed to be the primary celerybeat.
+    CELERYBEAT_LOCK_RETRY_TIME=90,
     # Resource manager worker name
-    RESOURCE_MANAGER_WORKER_NAME='resource_manager'
+    RESOURCE_MANAGER_WORKER_NAME='resource_manager',
+    # celerybeat constants
```
Comment (bmbouter) [inline tasking/pulp/tasking/constants.py]: s/celerybeat constants/The name of the celerybeat entry in the workers table/

File: tasking/pulp/tasking/constants.py [hunk 18]
```diff
@@ -4,6 +4,20 @@
...
+    SCHEDULER_WORKER_NAME = 'celerybeat',
```
Comment (bmbouter) [inline tasking/pulp/tasking/constants.py]: s/SCHEDULER/CELERYBEAT/

File: tasking/pulp/tasking/constants.py [hunk 19]
```diff
@@ -4,6 +4,20 @@
     # The interval in seconds for which the Celery Process monitor thread sleeps between
     # checking for missing Celery processes.
     CELERY_CHECK_INTERVAL=60,
+    # The interval in seconds during which the Celery Processes being monitored need to report a
+    # hearbeat to be considered active.
+    HEARTBEAT_MAX_AGE=200,
+    # The interval in seconds with which a secondary celerybeat tries to acquire the lock
+    # that is needed to be the primary celerybeat.
+    CELERYBEAT_LOCK_RETRY_TIME=90,
     # Resource manager worker name
```
Comment (bmbouter) [inline tasking/pulp/tasking/constants.py]: s/Resource manager worker name/The name of the resource manager entry in the workers table/
